assembly/index.ts

```diff
@@ -0,0 +1,47 @@
+// The entry file of your WebAssembly module.
+
+// memory is assumed:
+// [input][output]
+
+export function utf8ToUtf16(byteLength: i32, outputOffset: i32): i32 {
+  let inputOffset: i32 = 0;
+  while (inputOffset < byteLength) {
+    let byte1: u16 = load<u8>(inputOffset++);
+    if ((byte1 & 0x80) === 0) {
+      // 1 byte
+      store<u16>(outputOffset, byte1);
+      outputOffset += 2;
+    } else if ((byte1 & 0xe0) === 0xc0) {
+      // 2 bytes
+      let byte2: u16 = load<u8>(inputOffset++) & 0x3f;
+      // FIXME: consider endians
+      store<u16>(outputOffset, ((byte1 & 0x1f) << 6) | byte2);
+      outputOffset += 2;
+    } else if ((byte1 & 0xf0) === 0xe0) {
+      // 3 bytes
+      let byte2: u16 = load<u16>(inputOffset++) & 0x3f;
+      let byte3: u16 = load<u16>(inputOffset++) & 0x3f;
```

Comment:
Also it seems this should be:
```ts
      let byte2: u16 = load<u8>(inputOffset++) & 0x3f;
      let byte3: u16 = load<u8>(inputOffset++) & 0x3f;
```

Reply:
Right: eea2f04

Reply:
other hint which could potentially improve speed is using `u32` instead `u16` for every temporary variable.